ambry-store/src/main/java/com.github.ambry.store/LogSegment.java

```diff
@@ -0,0 +1,339 @@
+              + "] with end offset [" + getEndOffset() + "]");
+    }
+    if (position + buffer.remaining() > getEndOffset()) {
+      metrics.overflowReadError.inc();
+      throw new IllegalArgumentException("Cannot read from segment [" + file.getAbsolutePath() + "] from position [" +
+          position + "] for size [" + buffer.remaining() + "] because it exceeds the end offset [" + endOffset + "]");
+    }
+    long bytesRead = 0;
+    int size = buffer.remaining();
+    while (bytesRead < size) {
+      bytesRead += fileChannel.read(buffer, position + bytesRead);
+    }
+  }
+
+  /**
+   * Writes {@code size} number of bytes from the channel {@code channel} into the segment at {@code offset}.
+   * <p/>
+   * There is no state checking done for this function and the assumption is that the caller knows what is being done.
+   * <p/>
+   * The write is not started if it cannot be completed.
+   * @param channel The channel from which data needs to be written from.
+   * @param offset The offset in the segment at which to start writing.
+   * @param size The amount of data in bytes to be written from the channel.
+   * @throws IllegalArgumentException if {@code offset} < 0 or if there is not enough space for {@code offset } +
+   * {@code size} data.
+   * @throws IOException if data could not be written to the file because of I/O errors
+   *
+   */
+  void writeFrom(ReadableByteChannel channel, long offset, long size)
+      throws IOException {
+    if (offset < 0 || offset > capacityInBytes) {
+      throw new IllegalArgumentException(
+          "Provided offset [" + offset + "] is out of bounds for the segment [" + file.getAbsolutePath()
+              + "] with capacity [" + capacityInBytes + "]");
+    }
+    if (offset + size > capacityInBytes) {
+      metrics.overflowWriteError.inc();
+      throw new IllegalArgumentException("Cannot write to segment [" + file.getAbsolutePath() + "] from offset [" +
+          offset + "] for size [" + size + "] because it exceeds the capacity [" + capacityInBytes + "]");
+    }
+    long bytesWritten = 0;
+    while (bytesWritten < size) {
+      bytesWritten += fileChannel.transferFrom(channel, offset + bytesWritten, size - bytesWritten);
+    }
+    if (offset + size > endOffset.get()) {
+      endOffset.set(offset + size);
+    }
+  }
+
+  /**
+   * Gets the {@link File} and {@link FileChannel} backing this log segment. Also increments a ref count.
+   * <p/>
+   * The expectation is that a matching {@link #closeView()} will be called eventually to decrement the ref count.
+   * @param offset the offset that will used to start the read operation from.
+   * @return the {@link File} and {@link FileChannel} backing this log segment.
+   * @throws IllegalArgumentException if {@code offset} < 0 or is > {@link #getEndOffset()}.
+   */
+  Pair<File, FileChannel> getView(long offset) {
+    if (offset < 0 || offset > getEndOffset()) {
```

Comment:
I don't understand what `offset` will be passed to this method, but I think you want to do `offset >= getEndOffset()` here (for the reasons I mentioned earlier).